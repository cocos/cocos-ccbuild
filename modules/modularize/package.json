{
  "name": "@ccbuild/modularize",
  "version": "1.0.0",
  "description": "",
  "scripts": {
    "build": "tsc",
    "test": "jest",
    "api": "tsc && api-extractor run --local --verbose"
  },
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "files": [
    "./lib/**/*"
  ],
  "exports": {
    ".": {
      "types": "./src/index.ts",
      "default": "./lib/index.js"
    }
  },
  "keywords": [],
  "author": "PP",
  "license": "ISC",
  "dependencies": {
    "fs-extra": "~11.1.1",
    "path": "~0.12.7",
    "commander": "~11.0.0",
    "@ccbuild/utils": "workspace:*",
    "ejs": "~3.1.9",
    "glob": "~7.2.0"
  },
  "devDependencies": {
    "jest": "~29.5.0",
    "@types/node": "~20.3.3",
    "@types/fs-extra": "~11.0.1",
    "@types/ejs": "~3.1.2",
    "ts-jest": "~29.1.1",
    "@types/jest": "~29.5.2",
    "typescript": "~5.1.6",
    "@microsoft/api-extractor": "~7.36.0",
    "@types/glob": "~7.2.0"
  }
}
