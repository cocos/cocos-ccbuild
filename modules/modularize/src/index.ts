export * from './module-config';
export * from './module-query';
export * from './platform-config';