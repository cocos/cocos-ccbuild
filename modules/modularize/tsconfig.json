{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "./lib",
    "rootDir": "./src",
  },
  "include": [
    "./src/index.ts"
  ],
  "exclude": [
    "./test",
    "./lib/",
    "./static/"
  ]
}
