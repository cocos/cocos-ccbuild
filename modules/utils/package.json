{
  "name": "@ccbuild/utils",
  "version": "1.0.0",
  "description": "",
  "scripts": {
    "build": "tsc"
  },
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "files": [
    "./lib/**/*"
  ],
  "keywords": [],
  "author": "PP",
  "license": "ISC",
  "exports": {
    ".": {
      "types": "./src/index.ts",
      "require": "./lib/index.js",
      "default": "./lib/index.js"
    }
  },
  "dependencies": {
    "fs-extra": "~11.1.1",
    "path": "~0.12.7",
    "commander": "~11.0.0"
  },
  "peerDependencies": {
    "@ccbuild/modularize": "workspace:*"
  },
  "devDependencies": {
    "@types/node": "~20.3.3"
  }
}
