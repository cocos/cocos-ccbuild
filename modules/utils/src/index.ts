export * from './path';

export function asserts(expr: unknown, message?: string): boolean {
    if (!expr) {
        throw new Error(message);
    }
    return true;
}

export function isThenable (value: any) {
    // https://stackoverflow.com/a/53955664/10602525
    return Boolean(value && typeof value.then === 'function');
}