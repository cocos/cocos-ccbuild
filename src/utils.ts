export function asserts(expr: unknown, message?: string): boolean {
    if (!expr) {
        throw new Error(message);
    }
    return true;
}

export function filePathToModuleRequest (path: string) {
    return path.replace(/\\/g, '\\\\');
}