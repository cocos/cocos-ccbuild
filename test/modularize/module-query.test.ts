import { absolutePathFuncFactory, ps } from '@ccbuild/utils';
import { ModuleQuery } from '@ccbuild/modularize';

const absolutePath = absolutePathFuncFactory(__dirname);
const engineRoot = absolutePath('../test-engine-source');
function relativeToRoot (path: string): string {
    return ps.relative(engineRoot, path);
}

test('query test', async () => {
    const mq = new ModuleQuery({
        engine: engineRoot,
        platform: 'HTML5',
    });
    expect(await mq.getAllModules()).toMatchSnapshot('get all modules');
    expect(await mq.getExports('@module-query/env')).toMatchSnapshot('get @module-query/env exports');
    expect(await mq.getAllExports()).toMatchSnapshot('get all modules exports');
    expect(await mq.getConfig('@module-query/utils')).toMatchSnapshot('get config');
    const exportMap = await mq.getExportMap();
    for (const [k, v] of Object.entries(exportMap)) {
        exportMap[k] = ps.relative(engineRoot, v);
    }
    expect(exportMap).toMatchSnapshot('get export map');
});

test('resolve package json', async () => {
    const mq = new ModuleQuery({
        engine: engineRoot,
        platform: 'HTML5',
    });
    expect(relativeToRoot(await mq.resolvePackageJson('@module-query/env'))).toMatchSnapshot();
});

test('resolve platform export', async () => {
    const mq1 = new ModuleQuery({
        engine: engineRoot,
        platform: 'HTML5',
    });
    const resolved1 = await mq1.resolveExport('@module-query/env');
    if (resolved1) {
        expect(relativeToRoot(resolved1)).toMatchSnapshot();
    }
    const mq2 = new ModuleQuery({
        engine: engineRoot,
        platform: 'WEB_DESKTOP',
    });
    const resolved2 = await mq2.resolveExport('@module-query/env');
    if (resolved2) {
        expect(relativeToRoot(resolved2)).toMatchSnapshot();
    }
    const mq3 = new ModuleQuery({
        engine: engineRoot,
        platform: 'ALIPAY',
    });
    const resolved3 = await mq3.resolveExport('@module-query/env');
    if (resolved3) {
        expect(relativeToRoot(resolved3)).toMatchSnapshot();
    }
    const mq4 = new ModuleQuery({
        engine: engineRoot,
        platform: 'VIVO',
    });
    const resolved4 = await mq4.resolveExport('@module-query/env');
    if (resolved4) {
        expect(relativeToRoot(resolved4)).toMatchSnapshot();
    }
    const mq5 = new ModuleQuery({
        engine: engineRoot,
        platform: 'NATIVE',
    });
    const resolved5 = await mq5.resolveExport('@module-query/env');
    if (resolved5) {
        expect(relativeToRoot(resolved5)).toMatchSnapshot();
    }
    const mq6 = new ModuleQuery({
        engine: engineRoot,
        platform: 'ANDROID',
    });
    const resolved6 = await mq6.resolveExport('@module-query/env');
    if (resolved6) {
        expect(relativeToRoot(resolved6)).toMatchSnapshot();
    }
});

test('resolve types export', async () => {
    const mq = new ModuleQuery({
        engine: engineRoot,
        platform: 'HTML5',
    });
    const resolved = await mq.resolveExport('@module-query/utils');
    if (resolved) {
        expect(relativeToRoot(resolved)).toMatchSnapshot();
    }
});

test('resolve custom export', async () => {
    const mq = new ModuleQuery({
        engine: engineRoot,
        platform: 'HTML5',
        customExportConditions: ['custom2', 'custom1'],
    });
    const resolved = await mq.resolveExport('@module-query/env');
    if (resolved) {
        expect(relativeToRoot(resolved)).toMatchSnapshot();
    }
});

test('resolve editor and internal export', async () => {
    const mq = new ModuleQuery({
        engine: engineRoot,
        platform: 'ALIPAY',
    });
    const resolved1 = await mq.resolveExport('@module-query/env/editor');
    if (resolved1) {
        expect(relativeToRoot(resolved1)).toMatchSnapshot();
    }
    const resolved2 = await mq.resolveExport('@module-query/env/internal');
    if (resolved2) {
        expect(relativeToRoot(resolved2)).toMatchSnapshot();
    }
    const mq2 = new ModuleQuery({
        engine: engineRoot,
        platform: 'WEB_DESKTOP',
    });
    const resolved3 = await mq2.resolveExport('@module-query/env/internal');
    if (resolved3) {
        expect(relativeToRoot(resolved3)).toMatchSnapshot();
    }
    expect(await mq2.resolveExport('@module-query/env/test')).toBeUndefined();
    expect(await mq2.resolveExport('test')).toBeUndefined();
    expect(await mq2.resolveExport('./test')).toBeUndefined();
});
