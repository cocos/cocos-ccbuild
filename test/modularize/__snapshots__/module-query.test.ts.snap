// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query test: get @module-query/env exports 1`] = `
[
  "@module-query/env",
  "@module-query/env/editor",
  "@module-query/env/internal",
]
`;

exports[`query test: get all modules 1`] = `
[
  "@types/cc-ambient-types",
  "@module-query/env",
  "@module-query/utils",
]
`;

exports[`query test: get all modules exports 1`] = `
[
  "@types/cc-ambient-types",
  "@types/cc-ambient-types/query",
  "@module-query/env",
  "@module-query/env/editor",
  "@module-query/env/internal",
  "@module-query/utils",
]
`;

exports[`query test: get config 1`] = `
{
  "$schema": "../../../../../modules/modularize/lib/module-config.schema.json",
  "description": "",
  "exports": {
    ".": {
      "node": "./package.json",
      "types": "./src/index.ts",
    },
  },
  "license": "MIT",
  "name": "@module-query/utils",
  "scripts": {},
  "version": "1.0.0",
}
`;

exports[`query test: get export map 1`] = `
{
  "@module-query/env": "packages/@module-query/env/src/web/env.ts",
  "@module-query/env/editor": "packages/@module-query/env/src/editor-export/env.ts",
  "@module-query/env/internal": "packages/@module-query/env/src/internal-export/env.ts",
  "@module-query/utils": "packages/@module-query/utils/src/index.ts",
  "@types/cc-ambient-types": "@types/cc-ambient-types/src/index.d.ts",
  "@types/cc-ambient-types/query": "@types/cc-ambient-types/scripts/query.js",
}
`;

exports[`resolve custom export 1`] = `"packages/@module-query/env/src/custom/env2.ts"`;

exports[`resolve editor and internal export 1`] = `"packages/@module-query/env/src/editor-export/env.ts"`;

exports[`resolve editor and internal export 2`] = `"packages/@module-query/env/src/internal-minigame-export/env.ts"`;

exports[`resolve editor and internal export 3`] = `"packages/@module-query/env/src/internal-export/env.ts"`;

exports[`resolve package json 1`] = `"packages/@module-query/env/package.json"`;

exports[`resolve platform export 1`] = `"packages/@module-query/env/src/web/env.ts"`;

exports[`resolve platform export 2`] = `"packages/@module-query/env/src/web/env.ts"`;

exports[`resolve platform export 3`] = `"packages/@module-query/env/src/minigame/env.ts"`;

exports[`resolve platform export 4`] = `"packages/@module-query/env/src/runtime/env.ts"`;

exports[`resolve platform export 5`] = `"packages/@module-query/env/src/native/env.ts"`;

exports[`resolve platform export 6`] = `"packages/@module-query/env/src/native/env.ts"`;

exports[`resolve types export 1`] = `"packages/@module-query/utils/src/index.ts"`;
